src/main/java/org/apache/tomcat/jakartaee/Migration.java

```diff
@@ -415,14 +432,61 @@ private boolean migrateStream(String name, InputStream src, OutputStream dest) t
             Util.copy(src, dest);
             logger.log(Level.INFO, sm.getString("migration.skip", name));
         } else if (isArchive(name)) {
-            if (zipInMemory) {
-                logger.log(Level.INFO, sm.getString("migration.archive.memory", name));
-                convertedStream = migrateArchiveInMemory(src, dest);
-                logger.log(Level.INFO, sm.getString("migration.archive.complete", name));
-            } else {
-                logger.log(Level.INFO, sm.getString("migration.archive.stream", name));
-                convertedStream = migrateArchiveStreaming(src, dest);
-                logger.log(Level.INFO, sm.getString("migration.archive.complete", name));
+            // Only cache nested archives (e.g., JARs inside WARs), not top-level files
+            // Top-level files will have absolute paths starting with "/"
+            boolean isNestedArchive = !name.startsWith("/") && !name.startsWith("\\");
+
+            CacheEntry cacheEntry = null;
+            if (isNestedArchive && cache != null && cache.isEnabled()) {
+                // Buffer source to compute hash and check cache
+                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+                IOUtils.copy(src, buffer);
+                byte[] sourceBytes = buffer.toByteArray();
+
+                // Get cache entry (computes hash and marks as accessed)
+                cacheEntry = cache.getCacheEntry(sourceBytes);
+
+                if (cacheEntry.exists()) {
+                    // Cache hit! Copy cached result to dest and return
```

Comment:
Currently the cache is keyed solely on the hash of the input archive. It also needs to include the profile as the output may vary depending on the conversion profile being used.